### DISCUSSION TITLE
Search Form is not working

### DISCUSSION THREAD

**Discussion by liebeckm** [neuefische/questions#578] [Web] [answered]
Hey @ all,
I am currently working on my app for carpooling. I added a search form component that should browse through both, the rides that are currently hardcoded and the ones that the user can add manually. once the search button got clicked, the form should disappear and the results should be shown. currently, there are two problems:

1. the form gets displayed above the list of all rides
2. once I click the search button, the following error message occurs for "pages/search-rides/index.js":  "TypeError: Cannot read properties of undefined (reading 'departure')"

Can someone suggest a solution on this? :) 

Here is the needed information for you:

- [User Story](https://github.com/liebeckm/capstone-project/issues/6)
- [Search Form Component](https://github.com/liebeckm/capstone-project/blob/add-search/components/Form/RideSearchForm.styled.js)
- [Search Page](https://github.com/liebeckm/capstone-project/blob/add-search/pages/search-rides/index.js)
- [Hard Coded Lib with rides](https://github.com/liebeckm/capstone-project/blob/add-search/lib/rides.js)
- [Add Ride Component](https://github.com/liebeckm/capstone-project/blob/add-search/components/Form/RideAddForm.styled.js) and
- [Add Ride Page](https://github.com/liebeckm/capstone-project/blob/add-search/pages/add-rides/index.js), for better understanding where the other existing rides come from

**Comment by augustinanders** [comment]
Hi @liebeckm,

I took a look at your code and while I don't understand every detail yet, I noticed an issue with your ___`pages/search-rides/index.js`___ file. Specifically, the `handleSearch` function requires a `formData` prop.

Here's the function in question:

```
  const handleSearch = (formData) => {
    const filteredRides = initalRides.filter(
      (ride) =>
        ride.departure.toLowerCase() === formData.departure.toLowerCase() &&
        ride.destination.toLowerCase() === formData.destination.toLowerCase() &&
        ride.date === formData.date &&
        ride.time === formData.time &&
        ride.seats >= formData.seats
    );
    setFilteredRides(filteredRides);
    setIsSearchFormVisible(false);
  };
  ```
When you call the function without passing a prop to it (`<RideSearchForm onSearch={handleSearch} />`), `formData` remains undefined (you can test this with `console.log(formData)`). I believe this may be causing the error you're experiencing.

Here is an example how to give props to a onSearch function:
```
RideSearchForm onSearch={()=>{handleSearch(props)}}
```

**Comment by liebeckm** [comment]
Hey @augustinanders thank you for the reply. unfortunately, when I insert `RideSearchForm onSearch={()=>{handleSearch(props)}}
` into my code, it's not working :( Did I miss another needed change?

**Reply by augustinanders** [reply]
Did you insert the `FormData` as props? I don't know where your FormData comes from, so I can't help you there atm.

**Comment by onemanwenttomow** [comment, marked as answer]
Hello @liebeckm 

I think the issue might be that we are trying to filter the rides in both the ```RideSearchForm ``` and ```search-rides/index.js```

I think we can remove the filtering from the ```handleSearch``` function in ```RideSearchForm ``` and instead provide an object with all of the properties that you can passing to ```props.onSearch```. (I posted an alternative idea at the end of this comment).

``` js
  function handleSearch(e) {
    e.preventDefault();
    // const filteredRides =
    //   props.rides &&
    //   props.rides.filter((ride) => {
    //     return (
    //       ride.departure.toLowerCase() === departure.toLowerCase() &&
    //       ride.destination.toLowerCase() === destination.toLowerCase() &&
    //       ride.date === date &&
    //       ride.time === time &&
    //       ride.seats >= seats
    //     );
    //   });
    props.onSearch({ departure, destination, date, time, seats }); // here with create the new object and pass it to onSearch
  }
```

I think we might also want to take a look at the data in ```initalRides```. 

At the moment the dates look like this
```"28.05.2023"``` where as the dates from your form look like this ```"2023-05-28"```. This means that your current filter will not find any matching dates. 

Also i think we will have the same issue with the times. In ```initialRides``` the times look like this ```"16:00 Uhr"``` and in the data coming back from the form they look like this ```"16:00"```. I would consider removing the " Uhr" from the dummy data. 

Additional thoughts: 

* I think we should consider if we need ```RideSearchForm``` to be a controlled component. I think we could simplify things by making it an uncontrolled component ([see React docs](https://react.dev/learn/sharing-state-between-components#controlled-and-uncontrolled-components))
* We might also want to consider destructuring the props in ```RideSearchForm```. 
``` js 
// export default function RideSearchForm(props) {
export default function RideSearchForm({onSearch}) {

```



Hopefully this helps

**Comment by liebeckm** [comment]
Hey @onemanwenttomow luckily, there are the first search results shown, with the following conditions:
1. changed date and time as you suggested
2. deleted doubled filter function as you suggested
3. all fields have to be filled with information

Now, I want to make the form work, only with departure and destination filled in. Furthermore, I'd like to have the list of all rides hidden and only the results shown, once the search button is clicked. And the styling of date and time is not really user friendly, Tried to change this for the add form, but don't know, how to take this up for the search form. Lots of questions, sorry! :D I am a bit lost here

**Reply by onemanwenttomow** [reply]
@liebeckm Glad to hear the suggestions worked! Please check my previous response with "Mark as answer", so that others know this question is answered. 

In general it is best to reply in threads to the responses (i.e. in the "Write a reply" part). Otherwise every comment you write in the main thread might be interpreted as a possible answer. 

Also, i think these follow questions would be best served in another discussion thread so that we can keep the discussion board focussed and easier to search. 

That being said, here are some possible ideas for you to take a look at for the follow up questions. If you get stuck with any of these suggestions please open a new discussion. 

> Now, I want to make the form work, only with departure and destination filled in

You could remove the date, time and seats comparison from the filter inside of ```handleSearch```. That way you are only filtering based on departure and destination. 

> I'd like to have the list of all rides hidden and only the results shown, once the search button is clicked

You could set the initial value of rides to be an empty array. And then inside ```handleSearch``` you could filter over ```initialRides```

> And the styling of date and time is not really user friendly.
* You could look into the [Intl.DateTimeFormat() Constructor](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/DateTimeFormat/DateTimeFormat)
* If this does not cover the style of the date and time you are looking for then you may want to reach for a library (i.e. [date-fns](https://date-fns.org/))

Hopefully this helps.
